### PR TITLE
Datalake registered location

### LINKED ISSUE #7587: AWS Lakeformation registered location for S3 and role
### Describe the feature

We're trying to come up with ways to prevent data exfiltration via AWS LakeFormation. As per my understanding, the service lets you register an Amazon S3 path for data lake. The goal is to make sure that bucket is not an external bucket. 

Proposed solution:

Since we do not have a way to detect ownership of the bucket, we're trying to lock down the buckets to a list of approved ones. AWS allows usage of `prefix/` in the bucket path for finer grained access control. 
Secondly, we're also trying to look at the IAM role used to register the bucket. Potentially creating custom roles and locking them to our org id as per this new iam feature release (https://aws.amazon.com/blogs/security/how-to-control-access-to-aws-resources-based-on-aws-account-ou-or-organization/)

We could potentially add a `datalake-registered-location` resource type. The api gives back the following details.
`>>> response = client.list_resources()
>>> response
{'ResponseMetadata': {'RequestId': '77fe1b0b-5c79-43f1-81e9-1d648dcf4eea', 'HTTPStatusCode': 200, 'HTTPHeaders': {'date': 'Thu, 28 Jul 2022 22:37:34 GMT', 'content-type': 'application/json', 'content-length': '866', 'connection': 'keep-alive', 'x-amzn-requestid': '77fe1b0b-5c79-43f1-81e9-1d648dcf4eea', 'cache-control': 'no-cache'}, 'RetryAttempts': 0}, 'ResourceInfoList': [{'ResourceArn': 'arn:aws:s3:::test-1', 'RoleArn': 'arn:aws:iam::123456789:role/aws-service-role/lakeformation.amazonaws.com/AWSServiceRoleForLakeFormationDataAccess', 'LastModified': datetime.datetime(2022, 7, 26, 16, 58, 36, 160000, tzinfo=tzlocal())}, {'ResourceArn': 'arn:aws:s3:::test-4', 'RoleArn': 'arn:aws:iam::123456789:role/aws-service-role/lakeformation.amazonaws.com/AWSServiceRoleForLakeFormationDataAccess', 'LastModified': datetime.datetime(2022, 5, 3, 7, 56, 11, 572000, tzinfo=tzlocal())}, {'ResourceArn': 'arn:aws:s3:::test-2', 'RoleArn': 'arn:aws:iam::123456789:role/aws-service-role/lakeformation.amazonaws.com/AWSServiceRoleForLakeFormationDataAccess', 'LastModified': datetime.datetime(2022, 5, 3, 11, 4, 0, 568000, tzinfo=tzlocal())}, {'ResourceArn': 'arn:aws:s3:::test-3', 'RoleArn': 'arn:aws:iam::123456789:role/aws-service-role/lakeformation.amazonaws.com/AWSServiceRoleForLakeFormationDataAccess', 'LastModified': datetime.datetime(2022, 7, 26, 16, 57, 6, 196000, tzinfo=tzlocal())}]}`

A value filter to check the `RoleArn` and a filter to potentially look at S3 bucket name?

image

### Extra information or context

_No response_

### PR DESCRIPTION
**Pull request by PratMis**
Closes #7587 #7219 
For action to be taken i.e. deregister a location i made it a new resource type. Right after deregistering a location, it's expected to fix the service control role. I'm catching the below mentioned error here
```
botocore.errorfactory.InvalidInputException: An error occurred (InvalidInputException) when calling the DeregisterResource operation: Must manually delete service-linked role to deregister last S3 location.
```

Still need to do tests.

### CODE REVIEW

**Comment by kapilt** [comment]
do we want to have the cross-account filter here too then, to be able to use deregister in conjunction?

**Comment by PratMis** [comment]
> do we want to have the cross-account filter here too then, to be able to use deregister in conjunction?

Absolutely, that makes sense to me. I'll go ahead and add it along with tests. Thanks a lot for the feedback

**Comment by PratMis** [comment]
[update] AWS blocks usage of cross account role by default
```
botocore.errorfactory.AccessDeniedException: An error occurred (AccessDeniedException) when calling the RegisterResource operation: Cross-account pass role is not allowed.
```
I also looked at defining a class and using it for both `account` and `datalake-location` resource types. I guess that would also be a valid approach

File: c7n/resources/lakeformation.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                results.append(r)
+        return results
+
+    def process_account(self, r):
+        lake_bucket = {Arn.parse(r.get('ResourceArn')).resource}
+        buckets = {
+            b['Name'] for b in
+            self.manager.get_resource_manager('s3').resources(augment=False)}
+        cross_account = lake_bucket.difference(buckets)
+        if not cross_account:
```
Comment (thisisshi) [inline c7n/resources/lakeformation.py:67]: we can condense this to 

```python
return cross_account
```

**Review by thisisshi — APPROVED** [review]

**Pull request merged**